client/structs/funcs.go

```diff
@@ -1,62 +1,86 @@
 package structs
 
 import (
+	"errors"
 	"sync"
 
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
-// AllocBroadcaster implements an allocation broadcast channel.
-// The zero value is a usable unbuffered channel.
+const (
+	// listenerCap is the capacity of the listener chans. Must be exactly 1
+	// to prevent Sends from blocking and allows them to pop old pending
+	// updates from the chan before enqueueing the latest update.
+	listenerCap = 1
+)
+
+var ErrAllocBroadcasterClosed = errors.New("alloc broadcaster closed")
+
+// AllocBroadcaster implements an allocation broadcast channel where each
+// listener receives allocation updates. Pending updates are dropped and
+// replaced by newer allocation updates, so listeners may not receive every
+// allocation update. However this ensures Sends never block and listeners only
+// receive the latest allocation update -- never a stale version.
 type AllocBroadcaster struct {
 	m         sync.Mutex
-	listeners map[int]chan<- *structs.Allocation // lazy init
+	alloc     *structs.Allocation
+	listeners map[int]chan *structs.Allocation // lazy init
 	nextId    int
-	capacity  int
 	closed    bool
 }
 
-// NewAllocBroadcaster returns a new AllocBroadcaster with the given capacity (0 means unbuffered).
-func NewAllocBroadcaster(n int) *AllocBroadcaster {
-	return &AllocBroadcaster{capacity: n}
+// NewAllocBroadcaster returns a new AllocBroadcaster with the given initial
+// allocation.
+func NewAllocBroadcaster(initial *structs.Allocation) *AllocBroadcaster {
+	return &AllocBroadcaster{
+		alloc: initial,
+	}
 }
 
-// AllocListener implements a listening endpoint for an allocation broadcast channel.
+// AllocListener implements a listening endpoint for an allocation broadcast
+// channel.
 type AllocListener struct {
 	// Ch receives the broadcast messages.
 	Ch <-chan *structs.Allocation
 	b  *AllocBroadcaster
 	id int
 }
 
-// Send broadcasts a message to the channel. Send returns whether the message
-// was sent to all channels.
-func (b *AllocBroadcaster) Send(v *structs.Allocation) bool {
+// Send broadcasts an allocation update. Any pending updates are replaced with
+// this version of the allocation to prevent blocking on slow receivers.
+func (b *AllocBroadcaster) Send(v *structs.Allocation) error {
 	b.m.Lock()
 	defer b.m.Unlock()
 	if b.closed {
-		return false
+		return ErrAllocBroadcasterClosed
 	}
-	sent := true
+
+	// Update alloc on broadcaster to send to newly created listeners
+	b.alloc = v
+
+	// Send alloc to already created listeners
 	for _, l := range b.listeners {
 		select {
 		case l <- v:
-		default:
-			sent = false
+		case <-l:
```

Comment:
Seems like this code deadlock

Reply:
Great catch. Fixing.

Reply:
After some reviewing and testing I don't believe this code deadlocks since all broadcaster interactions are serialized with locks. If a Send() received on `l` (L66), then it will always be able to send on `l` (L67).

It's more subtle than I'd like. However, I really like the behavior (dropping stale updates), and I'm not sure of a better way to implement it.